apps/src/storage/dataBrowser/dataVisualizer/CrossTabChart.jsx

```diff
@@ -1,146 +1,187 @@
+  const pivotedColumns = new Set(records.map(record => record[columnName]));
 
-  render() {
-    if (
-      !this.props.records ||
-      !this.props.selectedColumn1 ||
-      !this.props.selectedColumn2
-    ) {
-      return null;
+  // Count occurrences of each row/column pair
+  records.forEach(record => {
+    let key = record[rowName];
+    let value = record[columnName];
+    if (!countMap[key]) {
+      countMap[key] = {[rowName]: key};
+      pivotedColumns.forEach(column => (countMap[key][column] = 0));
     }
-    const {chartData, columns} = this.createPivotTable(
-      this.props.records,
-      this.props.selectedColumn1,
-      this.props.selectedColumn2
-    );
-    const numericValues = [];
+    countMap[key][value]++;
+  });
 
-    chartData.forEach(record => {
-      Object.entries(record).forEach(entry => {
-        let key = entry[0];
-        let value = entry[1];
-        if (typeof value === 'number' && key !== this.props.selectedColumn1) {
-          numericValues.push(value);
-        }
-      });
+  // Sort columns
+  let columns;
+  if (numericColumns.includes(columnName)) {
+    columns = [...pivotedColumns].sort(function(a, b) {
+      return a - b;
     });
+  } else {
+    columns = [...pivotedColumns].sort();
+  }
+  columns.unshift(rowName);
 
-    const min = Math.min(...numericValues);
-    const max = Math.max(...numericValues);
+  // Sort rows
+  let chartData = Object.values(countMap).sort((a, b) =>
+    a[rowName] > b[rowName] ? 1 : -1
+  );
+  return {chartData, columns};
+}
 
-    return (
-      <div id={CROSS_TAB_CHART_AREA}>
-        <h1 style={styles.title}>{this.props.chartTitle}</h1>
-        <table>
-          <tbody>
-            <tr>
-              {columns.map(column => (
-                <th key={column} style={styles.headerCell}>
-                  {column}
-                </th>
-              ))}
-            </tr>
-            {chartData.map((record, id) => (
-              <tr key={id}>
-                {columns.map(column => {
-                  const value = record[column];
-                  const color =
-                    column === this.props.selectedColumn1
-                      ? 'white'
-                      : this.getColorForValue(value, min, max);
-                  const cellStyle = {...styles.cell, backgroundColor: color};
-                  return (
-                    <td key={column} style={cellStyle}>
-                      {value}
-                    </td>
-                  );
-                })}
-              </tr>
-            ))}
-          </tbody>
-        </table>
-      </div>
-    );
+export function getColorForValue(value, min, max) {
+  if (typeof value !== 'number') {
+    return 'white';
   }
+  const lightest = 100;
+  const darkest = 56;
+  const lightness =
+    lightest - ((value - min) / (max - min)) * (lightest - darkest);
+  return `hsl(217, 89%, ${lightness}%)`;
 }
 
-export default CrossTabChart;
+const wrapperStyle = {
```

Comment:
Mostly just curious- why make these all individual objects instead of one `styles` object that contains them?

Reply:
I switched to this to make it easier to compose styles, below. (See `topCellStyle`, `leftCellStyle`, etc.)